src/networks/infinitesquarenetwork.jl

```diff
@@ -74,42 +91,50 @@ function Base.isapprox(A₁::InfiniteSquareNetwork, A₂::InfiniteSquareNetwork;
 end
 
 ## Rotations
-function Base.rotl90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}} # Rotations of matrix unit cells
-    return NWType(rotl90(rotl90.(unitcell(A))))
-end
-function Base.rotr90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}}
-    return NWType(rotr90(rotr90.(unitcell(A))))
-end
-function Base.rot180(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}}
-    return NWType(rot180(rot180.(unitcell(A))))
-end
-function Base.rotl90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}} # Rotations of cubic unit cells along z-axis
-    return NWType(stack(rotl90, eachslice(unitcell(A); dims=3)))
+
+function Base.rotl90(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rotl90(_rotl90_localsandwich.(unitcell(n))))
 end
-function Base.rotr90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}}
-    return NWType(stack(rotr90, eachslice(unitcell(A); dims=3)))
+function Base.rotr90(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rotr90(_rotr90_localsandwich.(unitcell(n))))
 end
-function Base.rot180(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}}
-    return NWType(stack(rot180, eachslice(unitcell(A); dims=3)))
+function Base.rot180(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rot180(_rot180_localsandwich.(unitcell(n))))
 end
 
-## OptimKit optimization compatibility
-function LinearAlgebra.rmul!(A::InfiniteSquareNetwork, α::Number) # Used in _scale during OptimKit.optimize
-    rmul!.(unitcell(A), α)
-    return A
+## Chainrules
+
+function ChainRulesCore.rrule(
+    ::typeof(Base.getindex), network::InfiniteSquareNetwork, args...
+)
+    O = network[args...]
+
+    function getindex_pullback(ΔO_)
+        ΔO = map(unthunk, ΔO_)
+        if ΔO isa Tangent
+            ΔO = ChainRulesCore.construct(typeof(O), ChainRulesCore.backing(ΔO))
+        end
+        Δnetwork = zerovector(network)
+        Δnetwork[args...] = ΔO
+        return NoTangent(), Δnetwork, NoTangent(), NoTangent()
+    end
+    return O, getindex_pullback
 end
-function LinearAlgebra.axpy!(
-    α::Number, A₁::InfiniteSquareNetwork, A₂::InfiniteSquareNetwork
-) # Used in _add during OptimKit.optimize
-    axpy!.(α, unitcell(A₁), unitcell(A₂))
-    return A₂
+
+function ChainRulesCore.rrule(::typeof(rotl90), network::InfiniteSquareNetwork)
+    network´ = rotl90(network)
+    function rotl90_pullback(Δnetwork_)
+        Δnetwork = unthunk(Δnetwork_)
+        return NoTangent(), rotr90(Δnetwork)
+    end
+    return network´, rotl90_pullback
 end
 
-## FiniteDifferences vectorization
```

Comment:
This might be convenient to keep around?

Reply:
I kept it around for `InfinitePEPS`, which is what this was actually implemented for in the first place (but since I repurposed the old supertype name for a new thing it shows up as removed...). Do you think it's useful for the contractible network type? I would assume we only ever optimize over a physicsy network. Of course I can also add it for contractible networks, but I don't immediately see a clear use case.

Reply:
fair enough